### PR TITLE
Agent loop v2: Planning & Task Management (part 1: refactoring)

### PR DESCRIPTION
**Pull request by Pwuts**
## Background
The so-called _agent loop_ of Auto-GPT currently lacks structural planning and task management. This impedes its long-term performance and also hampers the implementation of retrieval augmentation (see #3536).

## Changes & other to-do's
**Part 1 - Refactoring (#4799):**
- [x] Rename module `agent` -> `agents`
- [x] Check out LangChain's "Plan and execute" tools ([docs](https://python.langchain.com/docs/modules/agents/agent_types/plan_and_execute), [code](https://github.com/hwchase17/langchain/blob/master/langchain/experimental/plan_and_execute/))
- [x] Refactor `Agent` into `BaseAgent` and `Agent` with sensible structure

**Part 2 - Planning & Task Management (#5077):**
- [ ] Add planner agent
- [ ] Add executor agent (potential solution: specialize existing `Agent` implenmentation into an executor agent)
- [ ] Evaluate & tweak

### CODE REVIEW

**Comment by zachary-kaelan** [comment]
The [Plan-and-Solve Prompting](https://arxiv.org/pdf/2305.04091.pdf) paper they cite is some good reading and has an [accompanying GitHub repo](https://github.com/AGI-Edgerunners/Plan-and-Solve-Prompting).

File: tests/unit/test_prompt_generator.py [hunk 1]
```diff
@@ -99,54 +99,3 @@ def test_generate_prompt_string(config):
...
-    for constraint in constraints:
-        assert constraint in prompt_string
-    for command in commands:
-        assert command["name"] in prompt_string
-        for key, value in command["args"].items():
-            assert f'"{key}": "{value}"' in prompt_string
-    for resource in resources:
-        assert resource in prompt_string
-    for evaluation in evaluations:
-        assert evaluation in prompt_string
```
Comment (Pwuts) [inline tests/unit/test_prompt_generator.py]: This is an exact duplicate of the test above this one, so I removed it

**Comment by coditamar** [comment]
@CodiumAI-Agent please review

**Comment by Pwuts** [comment]
Re: https://github.com/Significant-Gravitas/Auto-GPT/pull/4799#issuecomment-1624844975
> **✨ Minimal and focused:** Yes, the PR is focused on refactoring the agent loop for better planning and task management. All changes are related to this theme.

I think there's a screw loose somewhere, because in no universe is this PR minimal and focused :P

**Comment by coditamar** [comment]
> Re: https://github.com/Significant-Gravitas/Auto-GPT/pull/4799#issuecomment-1624844975
> > **✨ Minimal and focused:** Yes, the PR is focused on refactoring the agent loop for better planning and task management. All changes are related to this theme.
> 
> I think there's a screw loose somewhere, because in no universe is this PR minimal and focused :P


Agree.
The review process includes a "compression" process, which causes this issue/bug. To be improved.
https://github.com/Codium-ai/pr-agent/blob/main/PR_COMPRESSION.md

File: autogpt/llm/providers/openai.py [hunk 2]
```diff
@@ -53,7 +53,7 @@
             name="gpt-4-0613",
             prompt_token_cost=0.03,
             completion_token_cost=0.06,
-            max_tokens=8192,
+            max_tokens=8191,
```
Comment (Pwuts) [inline autogpt/llm/providers/openai.py:56]: > **Note**

During testing, I encountered this error:
> openai.error.InvalidRequestError: This model's maximum context length is 8192 tokens. However, you requested 8192 tokens (815 in the messages, 7377 in the completion). Please reduce the length of the messages or completion.

File: autogpt/agents/base.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+        self.history = MessageHistory(
+            self.llm,
+            max_summary_tlength=summary_max_tlength or self.send_token_limit // 6,
+        )
+
+    def __next__(self) -> None:
+        """Runs the agent for one cycle using its `default_cycle_instruction`.
+
+        Raises `StopIteration` if the `cycle_budget` is reached.
```
Comment (Pwuts) [inline autogpt/agents/base.py]: The idea I had in mind for this was to allow the agent to generate output on every cycle, which can be used by the parent to monitor or even steer its activity.

File: autogpt/agents/base.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+            if current_tokens_used + tokens_to_add > self.send_token_limit:
+                logger.debug(f"Plugin response too long, skipping: {plugin_response}")
+                logger.debug(f"Plugins remaining at stop: {plugin_count - i}")
+                break
+            prompt.insert(
+                -1, message_to_add
+            )  # HACK: assumes cycle instruction to be at the end
+            current_tokens_used += tokens_to_add
+
+    def context_while_think(self) -> ContextManager:
```
Comment (Pwuts) [inline autogpt/agents/base.py]: Can be used to run progress indicators etc while waiting for the LLM's response

File: autogpt/llm/providers/openai.py (hunk 2 above)
Comment (collijk) [reply autogpt/llm/providers/openai.py:56]: I was also getting this on your branch.  I don't think this is the fix though. Getting back 7400 tokens seems like something is wrong with the API.

File: tests/unit/test_agent.py [hunk 5]
```diff
@@ -2,9 +2,10 @@
 
 
 def test_agent_initialization(agent: Agent):
-    assert agent.ai_name == "Base"
+    assert agent.ai_config.ai_name == "Base"
```
Comment (collijk) [inline tests/unit/test_agent.py:5]: Deleting this test is also reasonable.  It's not very useful and has a lot of churn.

File: autogpt/llm/providers/openai.py (hunk 2 above)
Comment (collijk) [reply autogpt/llm/providers/openai.py:56]: This should be addressed in a separate issue.

**Review by collijk — APPROVED** [review]
These changes have been thoroughly reviewed.

**Comment by collijk** [comment]
> These changes have been thoroughly reviewed.

@Pwuts may have some finishing touches and some user testing to do and may need sign off from someone else not in the weeds of this PR.

File: autogpt/llm/providers/openai.py (hunk 2 above)
Comment (Pwuts) [reply autogpt/llm/providers/openai.py:56]: * https://github.com/Significant-Gravitas/Auto-GPT/issues/4985

**Review by merwanehamadi — DISMISSED** [review]

File: autogpt/main.py [hunk 6]
```diff
@@ -186,19 +188,212 @@ def run_auto_gpt(
...
+
+    def signal_handler(signum: int, frame: Optional[FrameType]) -> None:
+        if agent.cycle_budget == 0 or agent.cycles_remaining == 0:
+            sys.exit()
+        else:
+            logger_.typewriter_log(
+                "Interrupt signal received. Stopping continuous command execution.",
+                Fore.RED,
+            )
+            agent.cycles_remaining = 0
```
Comment (Pwuts) [inline autogpt/main.py]: Doesn't seem to work properly:
>   Cycle budget: 5; remaining: 5
>   Creating chat completion with model gpt-3.5-turbo, temperature 0.0, max_tokens 1023

`Ctrl`+`C`

> Aborted!
> Press any key to continue...

Comment (cyrus-hawk) [reply autogpt/main.py]: fixed the issue on https://github.com/Significant-Gravitas/Auto-GPT/pull/4997

Comment (Pwuts) [reply autogpt/main.py]: Thank you! :)

**Pull request merged**